### ISSUE TITLE
[question] How to integrate conan 2.0 to Visual Studio 2022?

### ISSUE THREAD

**Issue by Todiq** [conan-io/conan#14220]
### What is your question?

Hello,

I have several questions regarding compilation within the Visual Studio IDE.
I know that there is supposed to be an [extension](https://github.com/conan-io/conan-vs-extension) to help on that part at some point.

I have a multilib package. To build all its libs, I also need to compile a binary beforehand, which is part of the same package. In fact, it is used to generate the header files required for the rest.

In my recipe, the ```build(self)``` method looks like this:
```python
def build(self):
	cmake = CMake(self)
	cmake.configure()
	with VirtualRunEnv(self).vars().apply():
		cmake.build()
```

This works marvelously with ```conan build``` and ```conan create``` (both on Linux and Windows with cmake, using gcc and msvc respectively), since it loads the environment automatically.

However, I have to mimick the same behaviour within Visual Studio 2022 which, as far as I know, is only capable of manipulating cmake directly. So, here are my questions:

- **Is it possible to use ```conan build``` instead of directly using cmake?**
- If not, **Is it possible to configure the IDE to run ```conan install``` on its own before cmake generation**?
- Otherwise, **Can we integrate the generated .ps1/.bat (created manually with ```conan install```) containing the env variables to the IDE?** (conanrunenv.ps1 in this case).

Thanks in advance.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @Todiq 

Thanks for your question.

However, I am afraid that I didn't understand it. If the project is a CMake project, then building it with VS2022 is already there, built-in, no need to do anything special, the ``conan new cmake_lib`` template will do it. The ``with VirtualRunEnv(self).vars().apply():`` looks like not necessary and probably not recommended. The necessary environment from ``tool_requires`` should be automatically applied in ``conanbuild`` (not runenv, the run environment is not intended for build time, but for later execution time)

If there is a binary that needs to be built first, in the same package, then somehow the CMake build scripts will manage to do that, but that is also independent of the compiler (VS2022) used?

The VirtualRunEnv, the generated .bat scripts are necessary when there are dependencies, but this is also not explain in the question.

>  I have to mimick the same behaviour within Visual Studio 2022 

You mean, using the ``MSBuildDeps`` and ``MSBuildToolchain`` integrations?

**Comment by Todiq** [comment]
Sorry, I indeed didn't make myself very clear. Plus, I am not familiar with VS at all. I'll try to do better

> However, I am afraid that I didn't understand it. If the project is a CMake project, then building it with VS2022 is already there, built-in, no need to do anything special, the `conan new cmake_lib` template will do it.

That's the main point. I ran ```conan new cmake_lib -d name=test -d version=1.0 -d requires=boost/1.81.0``` to create a test project. Boost v1.81.0 is already installed on my machine with the following settings:
```python
self.options["boost/*"].shared=True
self.options["boost/*"].without_python=False
self.options["boost/*"].zlib=False
```
I edited the conanfile of the template to match these to avoid a recompile.

It was built with the following profile:
```ini
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.version=193
os=Windows
```

And I now use that one for my own projects (directly edited the ```default``` profile):
```ini
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=17
compiler.runtime=dynamic
compiler.version=193
os=Windows

[conf]
tools.build:skip_test=True
tools.env.virtualenv:powershell=True
```

However, whenever I open the folder containing the newly created files with VS, I get the following error: 
```
Erreur		CMake Error at D:\Documents\CAST\test\CMakeLists.txt:4 (find_package):
  Could not find a package configuration file provided by "boost" with any of
  the following names:

    boostConfig.cmake
    boost-config.cmake

  Add the installation prefix of "boost" to CMAKE_PREFIX_PATH or set
  "boost_DIR" to a directory containing one of the above files.  If "boost"
  provides a separate development package or SDK, be sure it has been
  installed.	test	D:\Documents\CAST\test\CMakeLists.txt	4	
```

I think that is an intended behaviour, since we are supposed to run a ```conan install``` beforehand so that VS is able to load the cmake configuration properly, aren't we?

So, is there a way to ensure that a ```conan install``` is run by the IDE at the beginning so that these files would already exist?

**Comment by memsharded** [comment]
Thanks for the clarification, I think that I understand better now.

The current "canonical" flow for CMake + VS would be:

- ``conan install ...``
- ``cmake --preset conan-default``
- Open VS generated solution


I dislike the open CMake project from VS flow, maybe I am old school, but I think it is much better the explicit flow, and I see way more developers using this flow than opening CMake projects directly with VS, as that integration seemed to be traditionally a bit problematic or at least confusing in some cases.

> So, is there a way to ensure that a conan install is run by the IDE at the beginning so that these files would already exist?

We had a VS plugin that automatically ran Conan directly in the IDE. Unfortunately this plugin is not maintained at this moment. We will get back to it at some point (at the moment we are focused on other integrations like cmake-conan and CLion), but it might take some time.

The cmake-conan integration in https://github.com/conan-io/cmake-conan/tree/develop2, automatically runs Conan when ``cmake`` is called. If ``cmake`` is called automatically from the IDE, that might achieve the desired functionality. However I think that integration for 2.0 is still wip, not ready for production yet, and in general it is more complicated, yet another level of complicated abstraction to avoid just typing 1 extra ``conan install`` line (the ``cmake --preset`` would be standard to me, I don't count it as an extra line, because in my experience that is the default approach that should be done even without Conan)

**Comment by Todiq** [comment]
> The current "canonical" flow for CMake + VS would be:
> 
> * `conan install ...`
> * `cmake --preset conan-default`
> * Open VS generated solution

This means using the ```MSBuildDeps``` and ```MSBuildToolchain```, right? I am currently using the ```CmakeDeps``` and ```CMakeToolchain``` ones. Would you recommend adding a condition in my conanfiles in case of a build with VS, in order to directly manipulate solutions? I guess that would be easier for the developers using the IDE.<br/><br/><br/><br/>


Furthermore, I still have the following problem:

I have a target ```libA``` which is used by ```exeA```. ```exeA``` generates headers required to compile ```libB```. However, whenever I try to execute ```exeA``` (which has just been successfully compiled) in an intermediate target before building ```libB``` , I get an error about missing DLLs (that ```exeA``` cannot find ```libA```).

I understand that this error occurs because I build directly through the IDE, instead of doing a ```conan build```.
I edited the ```conanfile``` on your advice (#13710) to load ```VirtualRunEnv(self).vars().apply()```. I am forced to do this way because the exe is part of the package. Not a ```build_requirements()``` that could be added through ```self.tool_requires()```. However, while that fixed the issue with a ```conan build``` (works on Windows and Linux), that is no sufficient for the IDE part.

Is there any way that I could load, within the IDE, the files created by ```conan install``` in order to have the environment properly loaded and already containing the path to the DLLs that were used to build ```exeA```? Or shoud I wait for the VS extension to be ready? Please let me know if you need more details. Thanks in advance.

**Comment by memsharded** [comment]
> This means using the MSBuildDeps and MSBuildToolchain, right? I am currently using the CmakeDeps and CMakeToolchain ones. Would you recommend adding a condition in my conanfiles in case of a build with VS, in order to directly manipulate solutions? I guess that would be easier for the developers using the IDE.

No, no. This would be with ``CMakeDeps`` and ``CMakeToolchain``. As long as cmake is being called, the Conan generators are ``CMakeXXX``. The ``MSBuildXXX`` ones are exclusively for native VS solutions/projects, without any cmake involved at all.


> I have a target libA which is used by exeA. exeA generates headers required to compile libB. However, whenever I try to execute exeA (which has just been successfully compiled) in an intermediate target before building libB , I get an error about missing DLLs (that exeA cannot find libA).

I'd need some clarification. Those targets are different Conan packages, or they are all artifacts inside one single Conan package? Can you please clarify the relation between targets and Conan packages?

**Comment by Todiq** [comment]
> No, no. This would be with `CMakeDeps` and `CMakeToolchain`. As long as cmake is being called, the Conan generators are `CMakeXXX`. The `MSBuildXXX` ones are exclusively for native VS solutions/projects, without any cmake involved at all.

Then there is something I don't get. Is a solution supposed to be created with `CMakeDeps` and `CMakeToolchain`?

> I'd need some clarification. Those targets are different Conan packages, or they are all artifacts inside one single Conan package? Can you please clarify the relation between targets and Conan packages?

Sure. Let me try again with an example:

- I have a multilib package. Let's call it `pkg1`
- `pkg1` has dependencies (mentionned in the conanfile through `self.requires()`) on `pkg2` (one of my own) and on `boost`
- `pkg1` contains multiple cmake targets. One is `libA` which links with a lib of `pkg2` and some from `boost`
- Another one of `pkg1`'s targets is `exeA` which needs `libA` but also the same lib from `pkg2` than `libA` does and some others from `boost`
- To build a second lib (still within `pkg1`), called `libB`, I first need to build and run `exeA`
- `libA` compiles successfully and so does `exeA` straight afterwards
- I try to run `exeA` after building it  (**still during the same compilation process**) --> I get a linking error caused by missing DLLs, which are `libA`, the one from `pkg2` and the ones from `boost`

Adding `VirtualRunEnv(self).vars().apply()` to the `build()` method fixes the issue, but only when running a `conan build`:

```python
def build(self):
	cmake = CMake(self)
	cmake.configure()
	with VirtualRunEnv(self).vars().apply():
		cmake.build()
```

I would like to reach the equivalent behaviour, but through the VS IDE, knowing that it directly runs `cmake` to build the project instead of `conan build`, thus not able to load the environment properly. Is it possible to do so, somehow?

**Comment by memsharded** [comment]
Thanks for the clarification.

I think that possibly the best approach would be to decouple ``libB`` into its own package that ``tool_requires("exeA")`` package. It seems that too much is being doing inside ``pkg1``.

However, I understand that this might not be possible, in that case the strategy needs to be different. It is not possible to rely on anything that is done in the ``build()`` method, because that method is not executed when building from IDE or building directly calling ``cmake ...``

All the necessary preparation for the projects to run must happen in the ``generate()`` method.
In this case, I thing that possibly the easiest approach is to do an "import" of the dlls of the dependencies, something in the line (not tested):
```python
def generate(self):
      for mydep in ("pkg2", "boost"):
            dep = self.dependencies[dep]
            src = dep.cpp_info.bindir
            dst = os.path.join(self.build_folder, self.cpp.build.bindir)
            copy(self, "*.dll", src=src, dst=dst)
```
That will put a copy of the dlls on the place where the binaries are build, and the executable should be able to find them. Beware that if you remove the folder, that will dissapear and a new ``conan install`` will be necessary.

Please give this a try and let us know.

**Comment by Todiq** [comment]
I think I'll go this way then, thank you.
However, once the copy has been done, is it possible to start the exe from the `CMakeLists.txt`, the same way we link with the libraries (pkg::exe)? Or should I run like this:
```cmake
COMMAND path_to_executable option1 option2
```
But this would mean having conditions depending on the OS, since the extension of the executable would be different. Is there a cleaner way?

**Comment by Todiq** [comment]
@memsharded 

Any ideas regarding this? Thanks in advance.

**Comment by memsharded** [comment]
> However, once the copy has been done, is it possible to start the exe from the CMakeLists.txt, the same way we link with the libraries (pkg::exe)? Or should I run like this:

Yes, I think that is normal CMake:
```cmake
add_executable(creator creator.cxx)

# add the custom command to produce created.c
add_custom_command(
  OUTPUT ${PROJECT_BINARY_DIR}/created.c
  DEPENDS creator
  COMMAND creator ${PROJECT_BINARY_DIR}/created.c
  )
```

**Comment by Todiq** [comment]
Thank you for the clarification @memsharded 

Regarding this:
> The current "canonical" flow for CMake + VS would be:
> 
> * `conan install ...`
> * `cmake --preset conan-default`
> * Open VS generated solution

This works well and I am able to build in Release mode. However, when I try to compile in any other configuration, it looks like VS cannot find the headers of the requirements anymore. I think that would be related to the preset passed to `cmake --preset`, but I can't be sure. Do you have any ideas?<br><br>

Also, in the end, I decided to create a seperate package for my executable. I removed `with VirtualRunEnv(self).vars().apply():` from my recipes and use the pkg with `tool_requires`. While that works on Linux & Windows using `conan build`, it still won't with the `conan install ...` + `cmake --preset conan-default` flow. I am getting an error the second I try to use the exe I have copied using the `generate` method (error MSB8066 in VS). Do I still need to copy the necessary DLLs for them to work in that particular case? Thanks in advance.

**Comment by memsharded** [comment]
> This works well and I am able to build in Release mode. However, when I try to compile in any other configuration, it looks like VS cannot find the headers of the requirements anymore. I think that would be related to the preset passed to cmake --preset, but I can't be sure. Do you have any ideas?

You need to install the other dependencies as well, so:
```
conan install .  # default Release
conan install . -s build_type=Debug
...
```
Otherwise, they will not be there. Once you have the different dependencies installed, you can do the ``cmake --preset conan-default`` and you should be able to switch in the IDE the configuration and build in release and debug.

When you have executable tools in the dependencies ``tool_requires`` they are added to the PATH environment variable in the ``conanbuild.bat`` file. It is necessary to call that file, so things are in the PATH to be found, so calling ``conanbuild.bat`` before calling ``cmake --preset`` would be the way to go. We will explore in the future the possibility of defining environment directly in CMake, but at the moment, it is done via environment scripts .bat in Windows and .sh in Linux.

**Comment by Todiq** [comment]
Ok, that's clear. I also had to start the solution from the same shell I ran the conanbuild script (ps1/bat) into, otherwise it still would not work. Thanks a lot!

**Comment by memsharded** [comment]
Nice!

This is the ticket you want to track regarding possible future addition of env-vars to CMake: https://github.com/conan-io/conan/issues/14271

**Comment by schwaerz** [comment]
@memsharded Just had a first look into the CMake integration in Conan2. Looks really good :)

One thing did not work as expected, and as you explained this some lines above, would like to ask my question here first:

I called conan install two times, one time for the debug build, once for the release build.

Now when I call `cmake --preset conan-default` cmake complains about a duplicate preset named `conan-default`. Which is correct, as both the debug and the release build preset is called like that:

```
$ cmake --preset conan-default
CMake Error: Could not read presets from C:/repositories/repo-name:
Duplicate preset: "conan-default"
```

I wonder whether this currenlty (using conan 2.0.13) is a known issue or I may be doing something wrong?

**Comment by schwaerz** [comment]
Just tried renaming the debug build preset (name & configure-preset values) to something else. After that it worked.

**Comment by memsharded** [comment]
>  Just had a first look into the CMake integration in Conan2. Looks really good :)

Happy to hear that! :) 

> I wonder whether this currenlty (using conan 2.0.13) is a known issue or I may be doing something wrong?

I think that this shouldn't happen, and I am not aware of this issue. If you could please create a new ticket for it, with steps to reproduce, that would be great.

**Comment by schwaerz** [comment]
Sure thing :) Thanks for the quick answer.

**Comment by schwaerz** [comment]
Just for reference: https://github.com/conan-io/conan/issues/14979
